libs/community/langchain_community/callbacks/openai_info.py

```diff
@@ -210,19 +212,53 @@ def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
 
     def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
         """Collect token usage."""
-        if response.llm_output is None:
-            return None
-
-        if "token_usage" not in response.llm_output:
-            with self._lock:
-                self.successful_requests += 1
-            return None
-
-        # compute tokens and cost for this request
-        token_usage = response.llm_output["token_usage"]
-        completion_tokens = token_usage.get("completion_tokens", 0)
-        prompt_tokens = token_usage.get("prompt_tokens", 0)
-        model_name = standardize_model_name(response.llm_output.get("model_name", ""))
+        # Check for usage_metadata (langchain-core >= 0.2.2)
+        try:
+            generation = response.generations[0][0]
+        except IndexError:
+            generation = None
+        if isinstance(generation, ChatGeneration):
+            try:
+                message = generation.message
+                if isinstance(message, AIMessage):
+                    usage_metadata = message.usage_metadata
+                else:
+                    usage_metadata = None
+            except AttributeError:
+                usage_metadata = None
+        else:
+            usage_metadata = None
+        if usage_metadata:
+            token_usage = {"total_tokens": usage_metadata["total_tokens"]}
+            completion_tokens = usage_metadata["output_tokens"]
+            prompt_tokens = usage_metadata["input_tokens"]
+            if response.llm_output is None:
+                # model name (and therefore cost) is unavailable in
+                # streaming responses
+                model_name = ""
+                if isinstance(generation, ChatGenerationChunk):
+                    warnings.warn("Cost estimates unavailable in streaming mode.")
```

Comment:
will this be noisy?

Reply:
yea don't think this is needed. also why can't we estimate costs in streaming mode?